server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/test/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutesTest.java

```diff
@@ -1590,6 +1634,195 @@ void restoreShouldNotAppendMessagesToAnOtherUserMailbox() throws Exception {
+                .then()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .body("statusCode", is(400))
+                    .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
+                    .body("message", is(notNullValue()))
+                    .body("details", is(notNullValue()));
+            }
+
+            @Test
+            void exportShouldReturnBadRequestWhenShareeIsInvalid() {
+                given()
+                    .queryParam("action", "export")
+                    .queryParam("sharee", "share@to#me@")
+                    .body(MATCH_ALL_QUERY)
+                .when()
+                    .post(USER.asString())
+                .then()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .body("statusCode", is(400))
+                    .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
+                    .body("message", is(notNullValue()))
+                    .body("details", is(notNullValue()));
+            }
+        }
+
+        @Nested
+        class TaskGeneratingTest {
+
+            @Test
+            void exportShouldReturnATaskCreated() {
+                given()
+                    .queryParam("action", "export")
+                    .queryParam("sharee", "sharee@james.org")
+                    .body(MATCH_ALL_QUERY)
+                .when()
+                    .post(USER.asString())
+                .then()
+                    .statusCode(HttpStatus.CREATED_201)
+                    .body("taskId", notNullValue());
+            }
+
+            @Test
+            void restoreShouldProduceASuccessfulTask() {
+                vault.append(USER, DELETED_MESSAGE, new ByteArrayInputStream(CONTENT)).block();
+                vault.append(USER, DELETED_MESSAGE_2, new ByteArrayInputStream(CONTENT)).block();
+
+                String taskId =
+                    given()
+                        .queryParam("action", "export")
+                        .queryParam("sharee", USER_2.asString())
+                        .body(MATCH_ALL_QUERY)
+                    .when()
+                        .post(USER.asString())
+                        .jsonPath()
+                        .get("taskId");
+
+                given()
+                    .basePath(TasksRoutes.BASE)
+                .when()
+                    .get(taskId + "/await")
+                .then()
+                    .body("status", is("completed"))
+                    .body("taskId", is(taskId))
+                    .body("type", is(DeletedMessagesVaultExportTask.TYPE))
+                    .body("startedDate", is(notNullValue()))
+                    .body("submitDate", is(notNullValue()))
+                    .body("completedDate", is(notNullValue()));
+            }
+        }
+
+        @Nested
+        class SharingTest {
+
+            @Test
+            void exportShouldCallZipperToZipDeletedMessages() throws Exception {
```

Comment:
You are asserting the `HOW`

You should be reading the shared content and do some ZIP asserts on it to assert the result (and answer the WHAT question)

Reply:
This is the definition of zip():
```
DeletedMessageZipper.zip(Stream<DeletedMessageWithContent>, OutputStream);
```
parameters are one time used objects. I can capture inputs, but I can't inspect them after all

Reply:
Yes, but we should turn the OutputSteam (the right abstraction when you write) into an inputstream (the right abstraction when you read)

Cf

```
PipedInputStream in = new PipedInputStream()
PipedOutputStream out = new PipedOutputStream(in)

//on thread 1
blobStore.save(in);

//on thread 2
zipper.zip(messages, out);
out.close()
```